distributed/s3.py

```diff
@@ -26,148 +32,525 @@
+        secret : string (None)
+            if not anonymous, use this password, if specified
+        kwargs : other parameters for boto3 session
+        """
+        self.anon = anon
+        self.key = key
+        self.secret = secret
+        self.kwargs = kwargs
+        self.connect(anon, key, secret, kwargs)
+        self.dirs = {}
+        self.s3 = self.connect(anon, key, secret, kwargs)
+    
+    def connect(self, anon, key, secret, kwargs):
+        tok = tokenize(anon, key, secret, kwargs)
+        if tok not in self._conn:
+            logger.debug("Open S3 connection.  Anonymous: %s",
+                         self.anon)
+            if self.anon:
+                from botocore import UNSIGNED
+                from botocore.client import Config                
+                s3 = boto3.Session().client('s3',
+                         config=Config(signature_version=UNSIGNED))
+            else:
+                s3 = boto3.Session(self.key, self.secret,
+                                   **self.kwargs).client('s3')
+            self._conn[tok] = s3
+        return self._conn[tok]
+
+    def __getstate__(self):
+        d = self.__dict__.copy()
+        del d['s3']
+        logger.debug("Serialize with state: %s", d)
+        return d
+
+    def __setstate__(self, state):
+        self.__dict__.update(state)
+        self.connect(self.anon, self.key, self.secret, self.kwargs)
+
+    def open(self, path, mode='rb', block_size=4*1024**2):
+        """ Open a file for reading or writing
+
+        Parameters
+        ----------
+        path: string
+            Path of file on S3
+        mode: string
+            One of 'rb' or 'wb'
+        block_size: int
+            Size of data-node blocks if reading
+        """
+        if 'b' not in mode:
+            raise NotImplementedError("Text mode not supported, use mode='%s'"
+                    " and manage bytes" % (mode + 'b'))
+        return S3File(self, path, mode, block_size=block_size)
+
+    def _ls(self, path, refresh=False):
+        """ List files below path
+
+        Parameters
+        ----------
+        path : string/bytes
+            location at which to list files
+        detail : bool (=True)
+            if True, each list item is a dict of file properties;
+            otherwise, returns list of filenames
+        refresh : bool (=False)
+            if False, look in local cache for file details first
+        """
+        path = path.lstrip('s3://').lstrip('/')
+        bucket, key = split_path(path)
+        if bucket not in self.dirs or refresh:
+            if bucket == '':
+                # list of buckets
+                if self.anon:
+                    # cannot list buckets if not logged in
+                    return []
+                files = self.s3.list_buckets()['Buckets']
+                for f in files:
+                    f['Key'] = f['Name']
+                    f['Size'] = 0
+                    del f['Name']
```

Comment:
To make a user's buckets look like top-level directories. That normalises (to my mind) a path like s3://mybucket/path/file ; you can do a ls('/') and see apparent directories.